redash_toolbelt/examples/migrate.py

```diff
@@ -0,0 +1,313 @@
+            response.raise_for_status()
+
+
+def import_visualizations():
+    print("Importing visualizations...")
+
+    for query_id, new_query_id in meta['queries'].items():
+        query = api_request('/api/queries/{}'.format(query_id))
+        print("   importing visualizations of: {}".format(query_id))
+
+        for v in query['visualizations']:
+            if v['type'] == 'TABLE':
+                response = requests.get(DESTINATION + '/api/queries/{}'.format(
+                    new_query_id), headers=auth_headers(DESTINATION_API_KEY))
+                response.raise_for_status()
+
+                new_vis = response.json()['visualizations']
+                for new_v in new_vis:
+                    if new_v['type'] == 'TABLE':
+                        meta['visualizations'][v['id']] = new_v['id']
+            else:
+                if str(v['id']) in meta['visualizations']:
+                    continue
+
+                user = user_with_api_key(query['user']['id'])
+                data = {
+                    "name": v['name'],
+                    "description": v['description'],
+                    "options": v['options'],
+                    "type": v['type'],
+                    "query_id": new_query_id
+                }
+                response = requests.post(
+                    DESTINATION + '/api/visualizations', json=data, headers=auth_headers(user['api_key']))
+                response.raise_for_status()
+
+                meta['visualizations'][v['id']] = response.json()['id']
+
+
+def import_dashboards():
+    print("Importing dashboards...")
+
+    dashboards = get_paginated_resource(ORIGIN + '/api/dashboards', ORIGIN_API_KEY)
+
+    for dashboard in dashboards:
+        print("   importing: {}".format(dashboard['slug']))
+        d = api_request('/api/dashboards/{}'.format(dashboard['slug']))
+        data = {'name': d['name']}
+        user = user_with_api_key(d['user_id'])
+        response = requests.post(
+            DESTINATION + '/api/dashboards', json=data, headers=auth_headers(user['api_key']))
+        response.raise_for_status()
+
+        new_dashboard = response.json()
+        requests.post(
+            DESTINATION + '/api/dashboards/{}'.format(new_dashboard['id']), json={'is_draft': False}, headers=auth_headers(user['api_key']))
+
+        # recreate widget
+        for widget in d['widgets']:
+            data = {
+                'dashboard_id': new_dashboard['id'],
+                'options': widget['options'],
+                'width': widget['width'],
+                'text': widget['text'],
+                'visualization_id': None
+            }
+
+            if not isinstance(widget['options'], dict):
+                widget['options'] = {}
+
+            if 'visualization' in widget:
+                data['visualization_id'] = meta['visualizations'].get(
+                    str(widget['visualization']['id']))
```

Comment:
In the first execution, the id is still a number. It's only after loading state from the JSON file that it becomes a string.

We should either store it as a string from the get go, or address this when loading the json file.

Reply:
To do this, I'm changing the metadata format from an object to an array of objects.

```
"users": {
        "1": {
            "id": "",
            "email": "",
            "invite_link": "",
            "api_key": ""
        },
    }
```
becomes...
```
{
  "users": [
    {
      "origin_id": 1,
      "destination_id": 1,
      "id": "",
      "email": "",
      "invite_link": "",
      "api_key": ""
    }
  ]
}
```

JSON doesn't permit using integers as keys to an object.

Reply:
Fixed bfb65ec23cd7926c9fd600ad2aa2803c8c3fa6b7

Reply:
Since this fix is old I'm just updating to say that this PR now coerces all string integers to ints when reading from the `meta.json` file.